### PR TITLE
First draft updated Python quickstart

### PR DESCRIPTION
**Pull request by LisaFC**
First draft of an updated Python quickstart with updating the proto, using protoc, etc.

If this works we can use it as a template for other languages....

### CODE REVIEW

**Comment by LisaFC** [comment]
@hsaliak @zinuga @nathanielmanistaatgoogle  - can you point this at appropriate people?

**Comment by LisaFC** [comment]
Preview at http://lcarey0-l.dub.corp.google.com:4000/docs/quickstarts/python-quickstart.html, though the server is being weirdly slow.

**Comment by LisaFC** [comment]
Ping? I'd like to get this approved so we can use it as a template for the other languages and get them into the new site.

File: docs/quickstarts/python-quickstart.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+Or, to install it system wide:
+
+```sh
+  $ sudo pip install grpcio
+```
+
+If you're on Windows, make sure you installed the `pip.exe` component when you
+installed Python. Invoke as above but with `pip.exe` instead of `pip` (you may
+also need to invoke from a `cmd.exe` ran as administrator):
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: `run as` rather than `ran as`? Honest question; I'm not entirely sure myself.

File: docs/quickstarts/python-quickstart.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+```sh
+  $ pip.exe install grpcio
+```
+
+### Install gRPC tools
+
+Python's gRPC tools include the protocol buffer compiler `protoc` and the special plugin for generating server and client code from `.proto` gRPC service definitions. For the first part of our quickstart example, we've already generated the server and client stubs from [helloworld.proto](https://github.com/grpc/grpc/tree/{{ site.data.config.grpc_release_branch }}/examples/protos/helloworld.proto), but you'll need the tools for the rest of our quickstart, as well as later tutorials and your own projects.
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: I suspect this should be `generating gRPC server and client code from .proto service definitions`. Nominally `.proto` files and protocol buffers in the large "don't know" about gRPC; they just offer a generic plug-in system of which gRPC makes use.

File: docs/quickstarts/python-quickstart.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+pip install grpcio-tools
+
+```
+
+## Download the example
+
+You'll need a local copy of the example code to work through this quickstart. Download the example code from our Github repository:
+
+```sh
+  $ # Clone the repository to get the example code:
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: [I am flat-out against this](https://github.com/grpc/grpc/issues/7097). The example code is five or six files; cloning the repository means downloading not only the entire repository but also all content it ever had.

How much time to we have to find a better option here?

File: docs/quickstarts/python-quickstart.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+
+## Run a gRPC application
+
+From the `examples/python/helloworld` directory:
+
+1. Run the server
+
+   ```sh
+   $ python2.7 greeter_server.py &
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: The ampersand is what runs the server in the background and allows the user to continue using the terminal in which they launched the server. So either we should kill the ampersand or kill "In another terminal, " below. I'm not sure which; maybe @soltanmm or @hsaliak have a preference one way or another?

File: docs/quickstarts/python-quickstart.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+   ```sh
+   $ python2.7 greeter_client.py
+   ```
+
+Congratulations! You've just run a client-server application with gRPC.
+
+## Update a gRPC service
+
+Now let's look at how to update the application with an extra method on the server for the client to call. Our gRPC service is defined using protocol buffers: you can find out lots more about how this works in [What is gRPC?]() and [gRPC Basics: Python][]. For now all you need to know is that both the server and the client "stub" have a `SayHello` RPC method that takes a `HelloRequest` parameter from the client and returns a `HelloResponse` from the server, and that this method is defined like this:
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: Colon in second sentence should be semicolon?

Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: I don't think the antecedent of "this" is sufficiently established before "how this works". Is it too wordy to say "you can find out lots more about how service definitions in .proto files work in (links)"?

Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: If possible please write this text on several lines in the source markdown file; I think it will still render as an ordinary paragraph in the reader-facing form.

File: docs/quickstarts/python-quickstart.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+message HelloRequest {
+  string name = 1;
+}
+
+// The response message containing the greetings
+message HelloReply {
+  string message = 1;
+}
+```
+Let's update this so that the `Greeter` service has two methods. In your IDE or text editor [what should we recommend, Python people?], open `examples/proto/helloworld.proto` and update it with a new `SayHelloAgain` method, with the same request and response types:
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: [Not touching that with a ten-meter pole](https://en.wikipedia.org/wiki/Editor_war). Not going anywhere near that.

How about "... has two methods. Edit `examples/proto/helloworld.proto` and update it..."?

File: docs/quickstarts/python-quickstart.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+  string name = 1;
+}
+
+// The response message containing the greetings
+message HelloReply {
+  string message = 1;
+}
+```
+
+Don't forget to save the file!
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: I feel like I wouldn't need this, but maybe I'm not representative of the documents audience. What about putting it in parentheses?

File: docs/quickstarts/python-quickstart.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+Next we need to update the gRPC code used by our application to use the new service definition. From the `examples/python/helloworld` directory:
+
+```
+$ ./run_codegen.sh
+```
+
+This script invokes the protocol buffer compiler as:
+
+```
+$ protoc -I ../../protos --python_out=. --grpc_out=. --plugin=protoc-gen-grpc=&#96;which grpc&lowbar;python&lowbar;plugin&#96; ../../protos/helloworld.proto</pre>
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: This line needs to be updated; it should look something like `python -m grpc.tools.protoc -I../../protos --python_out=. --grpc_python_out=. ../../protos/helloworld.proto`.

Actually that may be exactly it.

File: docs/quickstarts/python-quickstart.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+## Update and run the application
+
+We now have new generated server and client code, but we still need to implement and call the new method in the human-written parts of our example application.
+
+### Update the server
+
+In the same directory, open `greeter_server.py`. Implement the new method like this:
+
+```
+class Greeter(helloworld_pb2.BetaGreeterServicer):
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: Drop "Beta" from this name, [migrating this part of this example from the Beta API to the GA API](https://github.com/grpc/grpc/issues/4977).

File: docs/quickstarts/python-quickstart.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+```
+
+### Update the client
+
+In the same directory, open `greeter_client.py`. Call the new method like this:
+
+```
+def run():
+  channel = implementations.insecure_channel('localhost', 50051)
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: [`channel = grpc.insecure_channel('localhost:50051')`](https://github.com/grpc/grpc/issues/4977)

File: docs/quickstarts/python-quickstart.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+  stub = helloworld_pb2.beta_create_Greeter_stub(channel)
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: [Drop "`beta_`"](https://github.com/grpc/grpc/issues/4977).

File: docs/quickstarts/python-quickstart.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+  response = stub.SayHello(helloworld_pb2.HelloRequest(name='you'), _TIMEOUT_SECONDS)
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: [Drop "`, _TIMEOUT_SECONDS`"](https://github.com/grpc/grpc/issues/4977).

File: docs/quickstarts/python-quickstart.md [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+  print("Greeter client received: " + response.message)
+  response = stub.SayHelloAgain(helloworld_pb2.HelloRequest(name='you'), _TIMEOUT_SECONDS)
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: [Also drop "`, _TIMEOUT_SECONDS`" here](https://github.com/grpc/grpc/issues/4977).

File: docs/quickstarts/python-quickstart.md [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+Just like we did before, from the `examples/python/helloworld` directory:
+
+1. Run the server
+
+   ```sh
+   $ python2.7 greeter_server.py &
+   ```
+
+2. In another terminal, run the client
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: Same either-one-but-not-both issue here with ampersand and "In another terminal".

File: docs/quickstarts/python-quickstart.md (hunk 4 above)
Comment (soltanmm) [inline docs/quickstarts/python-quickstart.md]: I don't think we're in the business of teaching users what the heck is going on when their terminal spews output from backgrounded jobs (if they're not familiar with backgrounding and SIGHUPs and blah blah blah), and I'd prefer us to not enter that business.

**Comment by nathanielmanistaatgoogle** [comment]
Looks mostly good despite the large number of line-by-line comments.

I consider [eliminating gRPC repository cloning from the example](https://github.com/grpc/grpc/issues/7097) steps the largest obstacle (and a currently unsolved one) at this time.

File: docs/quickstarts/python-quickstart.md (hunk 4 above)
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: So to be concrete I believe @soltanmm is advising "kill the ampersand and retain 'In another terminal'".

File: docs/quickstarts/python-quickstart.md [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+```sh
+  $ sudo pip install grpcio
+```
+
+If you're on Windows, make sure you installed the `pip.exe` component when you
+installed Python. Invoke as above but with `pip.exe` instead of `pip` (you may
+also need to invoke from a `cmd.exe` ran as administrator):
+
+```sh
+  $ pip.exe install grpcio
```
Comment (hsaliak) [inline docs/quickstarts/python-quickstart.md]: Does it make sense to showcase the use of virtualenv/pyvenv here? 
https://packaging.python.org/installing/#creating-virtual-environments

Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: It makes sense if we can pull it off, because I think there is a much-lower risk of some weirdness on the user's machine interfering with our instructions working if the user is using a virtual environment.

What might gum up our shifting these instructions to virtualenv is if we can't tell a good story with run_codegen.sh. We can if we treat the name of the virtual environment as not up to the user. I'd be comfortable with that.

File: docs/quickstarts/python-quickstart.md (hunk 3 above)
Comment (hsaliak) [inline docs/quickstarts/python-quickstart.md]: Probably 5-10 days.. here is an alternate proposal, we break out the examples into a grpc-examples into separate repos and submodule them into the gRPC repo. This probably needs cross language/runtime coordination but its cleaner, buildable and testable. There might be duplication of the protos..

File: docs/quickstarts/python-quickstart.md (hunk 4 above)
Comment (hsaliak) [inline docs/quickstarts/python-quickstart.md]: +1 for getting rid of the ampersand. 
Otherwise, we should also carry instructions to clean up, eg:
Cleanup:
1) ``fg``
2) Ctrl+C

File: docs/quickstarts/python-quickstart.md (hunk 3 above)
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: Before saying too much here let's take the discussion over to [its proper issue](https://github.com/grpc/grpc/issues/7097). I suspect Python won't be able to do its own thing and we'll have to get team-wide buy-in for the language-independent idea of "users shouldn't have to clone the development repository just to work through the examples".

File: docs/quickstarts/python-quickstart.md (hunk 6 above)
Comment (hsaliak) [inline docs/quickstarts/python-quickstart.md]: I would recommend to just keep it brief.. 
"Edit `examples/proto/helloworld.proto` and .." 
If we make recommendations, we may have to consider proto3 support in the IDEs or editors.

File: docs/quickstarts/python-quickstart.md (hunk 9 above)
Comment (hsaliak) [inline docs/quickstarts/python-quickstart.md]: Aye, I better get to it.

File: docs/quickstarts/python-quickstart.md [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+```
+
+### Run!
+
+Just like we did before, from the `examples/python/helloworld` directory:
+
+1. Run the server
+
+   ```sh
+   $ python2.7 greeter_server.py &
```
Comment (hsaliak) [inline docs/quickstarts/python-quickstart.md]: Just a note to drop the ampersand here and make this section consistent with the changes discussed above.

File: docs/quickstarts/python-quickstart.md (hunk 3 above)
Comment (soltanmm) [inline docs/quickstarts/python-quickstart.md]: Unfortunately Github doesn't support `git-archive`ing (which would ostensibly solve this)... because... Github. There might be a path to getting the file without fetching the entire repository, but unfortunately I've yet to see a git command that provides such functionality (because for whatever reason `git` doesn't seem capable of dissociating refs from downloading **everything** *except* when invoking `git remote show`). Alternatively, one could set the clone `--depth` option to `1` to at least avoid pulling the whole history.

File: docs/quickstarts/python-quickstart.md (hunk 1 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: I think both are OK but "run as" is probably better. Changed.

File: docs/quickstarts/python-quickstart.md (hunk 2 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Makes sense. Done.

File: docs/quickstarts/python-quickstart.md (hunk 4 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: The ampersand, she is dead. :)

File: docs/quickstarts/python-quickstart.md (hunk 5 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done, done, and I've broken the lines, hopefully that will work when I commit it...

File: docs/quickstarts/python-quickstart.md (hunk 6 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 8 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 9 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 10 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 11 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 12 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 13 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 14 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 16 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+installed Python. Invoke as above but with `pip.exe` instead of `pip` (you may
+also need to invoke from a `cmd.exe` run as administrator):
+
+```sh
+  $ pip.exe install grpcio
+```
+
+### Install gRPC tools
+
+Python's gRPC tools include the protocol buffer compiler `protoc` and the special plugin for generating server and client code from `.proto` service definitions. For the first part of our quickstart example, we've already generated the server and client stubs from [helloworld.proto](https://github.com/grpc/grpc/tree/{{ site.data.config.grpc_release_branch }}/examples/protos/helloworld.proto), but you'll need the tools for the rest of our quickstart, as well as later tutorials and your own projects.
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: Let's break up this super long single line as well.

File: docs/quickstarts/python-quickstart.md [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+
+To install gRPC tools, run:
+
+```sh
+pip install grpcio-tools
+
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: Unnecessary blank line?

File: docs/quickstarts/python-quickstart.md [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+```
+
+Don't forget to save the file!
+
+## Generate gRPC code
+
+Next we need to update the gRPC code used by our application to use the new service definition. From the `examples/python/helloworld` directory:
+
+```
+$ ./run_codegen.sh
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: After the integration of [this pull request](https://github.com/grpc/grpc/pull/7419), this line will be `python run_codegen.py`.

File: docs/quickstarts/python-quickstart.md [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+class Greeter(helloworld_pb2.GreeterServicer):
+
+  def SayHello(self, request, context):
+    return helloworld_pb2.HelloReply(message='Hello, %s!' % request.name)
+
+  def SayHelloAgain(self, request, context):
+    return helloworld_pb2.HelloReply(message='Hello again, %s!' % request.name)
+
+...
+
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: Consider dropping this blank line? Maybe it can stay if it's providing the effect that you want after rendering.

File: docs/quickstarts/python-quickstart.md [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+```
+
+### Update the client
+
+In the same directory, open `greeter_client.py`. Call the new method like this:
+
+```
+def run():
+  channel = grpc.insecure_channel('localhost:50051')
+  stub = helloworld_pb2.create_Greeter_stub(channel)
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: I was mistaken in the last round of comments; this should actually be `helloworld_pb2.GreeterStub(channel)`. [This pull request](https://github.com/grpc/grpc/pull/7419), during the drafting of which I actually ran all the changed code, serves as the source of truth.

File: docs/quickstarts/python-quickstart.md [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+## Download the example
+
+You'll need a local copy of the example code to work through this quickstart. Download the example code from our Github repository:
+
+```sh
+  $ # Clone the repository to get the example code:
+  $ git clone https://github.com/grpc/grpc
+  $ # Navigate to the "hello, world" Python example:
+  $ cd grpc/examples/python/helloworld
+  ```
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: Does this triple-tick need to be without whitespace on its left? When I look at [this rendering](https://github.com/LisaFC/grpc.github.io/blob/96acfeb2839c10c17b54febe61f84815f1f71c4e/docs/quickstarts/python-quickstart.md) the content looks fine but when I look at [this rendering](http://lcarey0-l.dub.corp.google.com:4000/docs/quickstarts/python-quickstart.html) this area of the page is all screwed up.

File: docs/quickstarts/python-quickstart.md (hunk 7 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 17 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 18 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Well spotted. Yes.

File: docs/quickstarts/python-quickstart.md (hunk 19 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 20 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 21 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 22 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: No, it should be without the whitespace, looks fine now.

File: docs/quickstarts/python-quickstart.md [hunk 23]
```diff
@@ -0,0 +1,210 @@
...
+
+<p class="lead">This guide gets you started with gRPC in Python with a simple working example.</p>
+
+<div id="toc"></div>
+
+## Before you begin
+
+### Prerequisites
+
+[versions of pip, any other tools, etc. go here]
```
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Is there anything we need to add here so users don't have ARGH WRONG VERSION OF THING crises before they've even started?

Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: `pip` at least version eight, but I think that's it.

Anything else, @soltanmm? I'm reluctant to say `virtualenv` without converting the whole tutorial to be `virtualenv`-based.

**Comment by nathanielmanistaatgoogle** [comment]
Looks nearly good to me; the "pip and other tools go here" work item is the only thing left.

As much as I'd love it I don't think we're going to get [a good solution to eliminating cloning this week](https://github.com/grpc/grpc/issues/7097).

File: docs/quickstarts/python-quickstart.md [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+```
+
+## Run a gRPC application
+
+From the `examples/python/helloworld` directory:
+
+1. Run the server
+
+   ```sh
+   $ python2.7 greeter_server.py
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: @kpayson64 is right that these appearances of `python2.7` risk incorrectly implying that gRPC Python only works with Python 2.7. All invocations of the Python interpreter in this tutorial should all just be `python` (without version suffix).

If anyone complains about gRPC Python not working with 2.7, 3.4, or later, that's a legitimate bug report.

If anyone complains about gRPC Python not working with earlier-than-2.7 or 3.0-through-3.3, that's an opportunity to tell them to update. :-P

File: docs/quickstarts/python-quickstart.md [hunk 25]
```diff
@@ -0,0 +1,212 @@
...
+
+## Generate gRPC code
+
+Next we need to update the gRPC code used by our application to use the new service definition. From the `examples/python/helloworld` directory:
+
+```
+$ python2.7 run_codegen.py
+```
+
+This script invokes the protocol buffer compiler as:
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: These words made more sense when `run_codegen.sh`'s contents were exactly the same as the following line. Now that we're providing `run_codegen.py`, the contents of which call a Python function rather than invoke the Python interpreter, I think this line of text should be changed to something like "This script has the same effect as:".

File: docs/quickstarts/python-quickstart.md [hunk 26]
```diff
@@ -0,0 +1,212 @@
...
+```
+
+### Run!
+
+Just like we did before, from the `examples/python/helloworld` directory:
+
+1. Run the server
+
+   ```sh
+   $ python2.7 greeter_server.py
```
Comment (hsaliak) [inline docs/quickstarts/python-quickstart.md]: We should replace all occurrences of python2.7 with python.

File: docs/quickstarts/python-quickstart.md (hunk 25 above)
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: Also, I had sort of been thinking of suggesting that we drop these two lines as a distraction, and @soltanmm pushed me further in that direction with his comment [here](https://github.com/grpc/grpc/pull/7419#discussion_r71219411).

File: docs/quickstarts/python-quickstart.md (hunk 23 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Added the `pip` version....

File: docs/quickstarts/python-quickstart.md (hunk 15 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Ok, should I hold off until we've decided whether we can do this or not?

Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: No, let's go forward without `virtualenv`. If a user doesn't yet know about `virtualenv` let's not take on the task of teaching them here.

File: docs/quickstarts/python-quickstart.md [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+<p class="lead">This guide gets you started with gRPC in Python with a simple working example.</p>
+
+<div id="toc"></div>
+
+## Before you begin
+
+### Prerequisites
+
+* `pip`: version 8 or higher
+* If you haven't done so already, install the gRPC runtime and gRPC tools
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: I don't think this line belongs here because how to make this two installations is the work of the next two `###`-delimited sections.

File: docs/quickstarts/python-quickstart.md (hunk 25 above)
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: After meditating on it further: let's rip this out. It's not necessary for completion of the example, it's a little bit of a distraction, and it's something that adept and curious users can work out how to do on their own if they want to author a command-line command that runs our code generator.

File: docs/quickstarts/python-quickstart.md (hunk 27 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 24 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md (hunk 25 above)
Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Done.

File: docs/quickstarts/python-quickstart.md [hunk 28]
```diff
@@ -0,0 +1,204 @@
...
+
+To install gRPC tools, run:
+
+```sh
+pip install grpcio-tools
+```
+
+## Download the example
+
+You'll need a local copy of the example code to work through this quickstart. Download the example code from our Github repository (the command clones the entire repo but you just need the examples for this quickstart and other tutorials):
```
Comment (nathanielmanistaatgoogle) [inline docs/quickstarts/python-quickstart.md]: "the following command" rather than just "the command"? And "repository" rather than "repo"?

Comment (LisaFC) [inline docs/quickstarts/python-quickstart.md]: Yes, sounds better. Done.

**Comment by nathanielmanistaatgoogle** [comment]
Luhguhtum!

**Comment by hsaliak** [comment]
So while this is a pull request to the master branch, it should go to the GA branch. Lisa would you be point this against the GA branch so that it can be merged?

**Comment by LisaFC** [comment]
Done: https://github.com/grpc/grpc.github.io/pull/259
